SwaprDAO/swapr-info
# fix/account fees

# Summary

Fixes #76 

Fix to correctly compute the accumulated fees for a specific account on the `Account` page. Currently, there are cases where the fee values are negative.
The changes improve the computation of the cumulative fees, but we should be aware that this solution is still not 100% accurate.

This PR also fixes a problem with the `Liquidity value` chart on the `Account` page, that was causing a missing render.

# How To Test
1.  Open the page `Accounts` and check that accounts with negative fees now show more reasonable values (between the deployed version and [dxstats](https://dxstats.eth.limo/)).

For example, these accounts were all displaying inconsistent negative values:
```
Gnosis Chain:
- 0x40b37ba95f9bcf8930e6f8a65e5b4534518c3eab
- 0x0dedd9ed281a4cb7fbc91a4ef7790dd8f669c899
- 0x4ffad6ac852c0af0aa301376f4c5dea3a928b120
```
```
Mainnet:
- 0x9e0394c971aebe048710e84dec74f3e94eb11f87
- 0xc6130400c1e3cd7b352db75055db9dd554e00ef0
- 0x48870f80f675eb9009e5c5b3a1a11bb58f836e2a
```
```
Arbitrum:
- 0x22f340c2604dc1cdbe26cac5838ea9ebc8862a46
- 0x5b948d87b002ebb79e77df0400961d41eaab402e
- 0x738671901affbc417fa370cf936e461bb5315bd0
```

# Background

Before these changes, the cumulative fees were computed by processing all the `liquidityPositionSnapshots` and `liquidityMiningPositionSnapshots` and calculating the earned fees between each position.
The issue was that when a user decided to stake, for example, all his LP tokens in a farming campaign 2 things happened:

1. A `liquidityPositionSnapshots` is created with a `liquidityTokenBalance = 0`
2. In the same timestamp a `liquidityMiningPositionSnapshots` is created with a `stakedLiquidityTokenBalance = <amount_staked>`

So processing these snapshots would cause problems since the user still has his LP tokens as `liquidityPosition` even though he staked them in a campaign.

With this PR we "sanitize" the `liquidityPositionSnapshots` by merging the `liquidityMiningPositionSnapshots` back into the liquidity position when the two happened in the same timestamp; this way the LP tokens are correctly still owned by the user in the `liquidityPositionSnapshot`. The merged `liquidityMiningPositionSnapshots` then can be discarded for the computation of the fees since we restored the `liquidityTokenBalance` on the `liquidityPositionSnapshot`.

For example, giving these snapshots:

```json
{
    "timestamp": 1634252275,
    "reserveUSD": "2107960.884030910187087076",
    "liquidityTokenBalance": "127.670703162190806133",
    "liquidityTokenTotalSupply": "16597.25568880324490369",
    "reserve0": "278.888100033622336734",
    "reserve1": "1053980.442015455093543538",
    "token0PriceUSD": "3779.223430072451175882440106077335",
    "token1PriceUSD": "1",
    "pair": {},
    "__typename": "LiquidityPositionSnapshot"
},
{
    "timestamp": 1634252985,
    "reserveUSD": "2162084.622183194269458128",
    "liquidityTokenBalance": "127.670703162190806133",
    "liquidityTokenTotalSupply": "2060.611164436867873828",
    "reserve0": "285.527435417365018316",
    "reserve1": "1081042.311091597134729064",
    "token0PriceUSD": "3786.124123278736317125479640965464",
    "token1PriceUSD": "1",
    "pair": {},
    "__typename": "LiquidityMiningPositionSnapshot"
},
{
    "timestamp": 1634252985,
    "reserveUSD": "2162084.622183194269458128",
    "liquidityTokenBalance": "0",
    "liquidityTokenTotalSupply": "16876.982978280333312153",
    "reserve0": "285.527435417365018316",
    "reserve1": "1081042.311091597134729064",
    "token0PriceUSD": "3786.124123278736317125479640965464",
    "token1PriceUSD": "1",
    "pair": {},
    "__typename": "LiquidityPositionSnapshot"
}
```

since the second and the third one have the same `timestamp` we move the `liquidityTokenBalance` of the `LiquidityMiningPositionSnapshot` back to the `LiquidityPositionSnapshot` in the same timestamp, obtaining:

```json
{
    "timestamp": 1634252275,
    "reserveUSD": "2107960.884030910187087076",
    "liquidityTokenBalance": "127.670703162190806133",
    "liquidityTokenTotalSupply": "16597.25568880324490369",
    "reserve0": "278.888100033622336734",
    "reserve1": "1053980.442015455093543538",
    "token0PriceUSD": "3779.223430072451175882440106077335",
    "token1PriceUSD": "1",
    "pair": {},
    "__typename": "LiquidityPositionSnapshot"
},
{
    "timestamp": 1634252985,
    "reserveUSD": "2162084.622183194269458128",
    "liquidityTokenBalance": "127.670703162190806133",
    "liquidityTokenTotalSupply": "16876.982978280333312153",
    "reserve0": "285.527435417365018316",
    "reserve1": "1081042.311091597134729064",
    "token0PriceUSD": "3786.124123278736317125479640965464",
    "token1PriceUSD": "1",
    "pair": {},
    "__typename": "LiquidityPositionSnapshot"
}
```
